File: chatapp/app/channels/status_channel.rb
class StatusChannel < ApplicationCable::Channel
  def subscribed
    stream_from "status_channel"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
    if count_connection?(current_user.id)
      ActionCable.server.broadcast "status_channel", {user_id: current_user.id, status: "offline"}
      @user = User.find_by(id: params[:user_id])
      @user.update(status: false)  
    end
  end

  def appear
    ActionCable.server.broadcast "status_channel", {user_id: current_user.id, status: "online"}
    @user = User.find_by(id: params[:user_id])
    @user.update(status: true)
  end

  def count_connection?(current_user)
    users_id = ActionCable.server.open_connections_statistics.map { |con| con[:subscriptions].map { |sub| JSON.parse(sub)["user_id"] } }.flatten
    !users_id.any?{ |user| user.to_i == current_user }
  end
end

File: chatapp/app/javascript/channels/status_channel.js
import consumer from "channels/consumer"

const data = document.getElementById("user-data")
const user_id = data.getAttribute("user-data-id")

consumer.subscriptions.create({channel: "StatusChannel", user_id: user_id}, {
  connected() {
    // Called when the subscription is ready for use on the server
    this.perform("appear")
  },

  disconnected() {
    // Called when the subscription has been terminated by the server
  },

  received(data) {
    // Called when there's incoming data on the websocket for this channel
    let elements = document.getElementsByClassName(`user-${data.user_id}-status`);
    window.elements = elements
    for (var i = 0; i < elements.length; i++){
      if(data.status == 'online'){
        elements[i].classList.add('online')
      } else{
        elements[i].classList.remove('online')
      }
    }
  }
});
